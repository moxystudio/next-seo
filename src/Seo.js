import React from 'react';
import PropTypes from 'prop-types';
import Head from 'next/head';

const commonMetaAttributes = ['name', 'http-equiv', 'property', 'itemprop'];
const commonLinkAttributes = ['rel', 'href'];

const getKey = (attributes, common) => {
    const key = common.map((attr) => attributes[attr] ? `${attr}_${attributes[attr]}` : '').filter(Boolean).join('#');

    return key || Math.random().toString(36).substr(2, 9);
};

const renderLink = (data) => data.map((attributes) =>
    <link key={ getKey(attributes, commonLinkAttributes) } { ...attributes } />,
);

const renderMeta = (data) => data.map((attributes) =>
    <meta key={ getKey(attributes, commonMetaAttributes) } { ...attributes } />,
);

const renderData = (data) => Object.entries(data).map(([key, value]) => {
    switch (key) {
    case 'title':
        return <title key="title">{ value }</title>;
    case 'meta':
        return renderMeta(value);
    case 'link':
        return renderLink(value);
    default:
        return null;
    }
});

const Seo = ({ data }) =>
    <Head>{ renderData(data) }</Head>;

Seo.propTypes = {
    data: PropTypes.shape({
        title: PropTypes.string,
        meta: PropTypes.arrayOf(PropTypes.object),
        link: PropTypes.arrayOf(PropTypes.object),
    }).isRequired,
};

export default Seo;
