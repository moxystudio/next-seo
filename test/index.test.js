import React from 'react';
import { shallow } from 'enzyme';
import Seo from '../src/Seo';

jest.mock('next/head', () => ({ children }) => <div data-test="mock-head">{ children }</div>);

const defaultProps = {
    data: {
        title: 'test title',
        meta: [
            {
                name: 'description',
                content: 'test description',
            },
            {
                property: 'og:title',
                content: 'test title',
            },
            {
                itemprop: 'name',
                content: 'test name',
            },
            {
                'http-equiv': 'content-type',
                content: '30',
            },
            {
                name: 'description',
                property: 'og:description',
                content: 'test description',
            },
            {
                foo: 'bar',
            },
        ],
        link: [
            {
                rel: 'icon',
                href: '/favicon.ico',
            },
            {
                rel: 'stylesheet',
                href: '/styles.css',
            },
        ],
        foo: 'bar',
    },
};

const renderWithProps = (props = {}) => shallow(<Seo { ...defaultProps } { ...props } />);

describe('Seo Component', () => {
    it('should render correctly with props', () => {
        const tree = renderWithProps();

        expect(tree).toMatchSnapshot();
    });
});
