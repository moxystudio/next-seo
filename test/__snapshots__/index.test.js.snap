// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Seo Component should render correctly with props 1`] = `
<Component>
  <title
    key="title"
  >
    test title
  </title>
  <meta
    content="test description"
    key="name_description"
    name="description"
  />
  <meta
    content="test title"
    key="property_og:title"
    property="og:title"
  />
  <meta
    content="test name"
    itemprop="name"
    key="itemprop_name"
  />
  <meta
    content="30"
    http-equiv="content-type"
    key="http-equiv_content-type"
  />
  <meta
    content="test description"
    key="name_description#property_og:description"
    name="description"
    property="og:description"
  />
  <meta
    foo="bar"
    key="4fzzzxrvu"
  />
  <link
    href="/favicon.ico"
    key="rel_icon#href_/favicon.ico"
    rel="icon"
  />
  <link
    href="/styles.css"
    key="rel_stylesheet#href_/styles.css"
    rel="stylesheet"
  />
</Component>
`;
